_posts/on-working-with-viewmodel-composition-based-system.md

```diff
@@ -0,0 +1,145 @@
+
+> This article won't be one of those "real world" kinds of articles that usually have nothing to do with the real world. It's near to impossible to represent the real world in an article. Maybe a book can aim to do it. Maybe.
+
+ViewModel Composition and UI Composition, or micro-frontends if you prefer, are challenging. They require a deep understanding of the architectural concepts, underlying technologies and technical stacks, and deployment pipelines and tools.
+
+> For a detailed description of ViewModel and UI Composition differences, have a look at my ["On ViewModel Composition and UI Composition"](https://milestone.topics.it/view-model-composition/2021/04/20/on-viewmodel-composition-and-ui-composition.html) article.
+
+Instead of modeling a real-world scenario, I'll use a trivial one that comes with all the challenges anyway. And hopefully, it'll be detailed enough to provide insights on designing a real system using ViewModel Composition techniques.
+
+## Architecture view models
+
+Imagine a trivial scenario where we're composing data from two services in one ViewModel to serve the frontend:
+
+<div class="mermaid">
+graph LR
+    subgraph Webserver A
+        A[[Rest API]]
+    end
+    subgraph Webserver B
+        B[[Rest API]]
+    end
+    subgraph Composition Gateway
+        A --> AH[[Composition Handler A]]
+        B --> BH[[Composition Handler B]]
+        AH --> VM[ViewModel]
+        BH --> VM[ViewModel]
+    end
+    subgraph Browser
+        VM-->F[[Frontend]]
+    end
+</div>
+
+If we're building a distributed system, we're probably also following, more or less intentionally, the [4+1 architecture view model](https://en.wikipedia.org/wiki/4%2B1_architectural_view_model).
+
+The presented diagram is a mix of the physical and process views. It's strictly neither of the two. However, it's useful for our discussion because, in some way, it's the final state: Whatever we're doing, the system, when deployed in production, needs to look like that.
+
+The logical view, however, is quite different:
+
+<div class="mermaid">
+graph LR
+    subgraph Logical service A
+        A((Data))
+        A --> AH[Composition Handler A]
+    end
+    subgraph Logical service B
+        B((Data))
+        B --> BH[Composition Handler B]
+    end
+    subgraph Branding service
+        AH --> VM[ViewModel]
+        BH --> VM
+        VM-->F[Presentation]
+    end
+</div>
+
+> I'm intentionally skipping UI Composition/micro-frontends. It can only be more complex. Please, leave a comment below if you think UI Composition aspects need any further in-depth analysis.
+
+In between the logical and physical/process view of the system sits the development view. The development view enables the logical view to get to life. At the same time, it mediates and adapts from the logical view to the runtime needs respecting the non-functional requirements imposed by the tool-chain in use.
```

Comment:
I switched "needs" to "requirements" because I read "needs" as a verb and couldn't follow the syntax. I don't really like using "requirements" twice in the same sentence but I think it's okay here.

```suggestion
In between the logical and physical/process view of the system sits the development view. The development view enables the logical view to come to life. At the same time, it mediates and adapts from the logical view to the runtime requirements respecting the non-functional requirements imposed by the tool-chain in use.
```